Changes to OpLog.h to adapt to kafka oplog code in the enterprise repo.

yige-hu opened the pull request:
Need to change OpTypes to plain enum (instead of enum class) so that it
can be implicitly casted to integers, since we want to use
optypes_enum2str[] for operation name printing.

ddn0 commented on scripts/check_format.sh:
```diff
@@ -26,7 +26,7 @@ while read -d '' filename; do
     fi
   fi
-done < <(find ${ROOTS} -name experimental -prune -o -name '*.cpp' -o -name '*.h' -print0)
+done < <(find ${ROOTS} -name experimental -prune -o -name '*.cpp' -print0 -o -name '*.h' -print0)
```
I can confirm that this version and the one previously work for me. And the new version seems more correct.

Out of curiosity, what version of find are you running?

```
ddn@spork:~/w/katana$ find --version
find (GNU findutils) 4.6.0.225-235f
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Eric B. Decker, James Youngman, and Kevin Dalley.
Features enabled: D_TYPE O_NOFOLLOW(enabled) LEAF_OPTIMISATION FTS(FTS_CWDFD) CBO(level=2)
```

tylershunt replied:
Here's my version:
```
$ find --version
find (GNU findutils) 4.7.0-git
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Eric B. Decker, James Youngman, and Kevin Dalley.
Features enabled: D_TYPE O_NOFOLLOW(enabled) LEAF_OPTIMISATION FTS(FTS_CWDFD) CBO(level=2)
```
Weird that they would change the behavior like that!
```
$ find ./libtsuba/src/tsuba_api.cpp -name *.cpp
./libtsuba/src/tsuba_api.cpp
$ find ./libtsuba/src/tsuba_api.cpp -name *.cpp -o -name *.o -print0
$
```

yige-hu committed:
Need to change OpTypes to plain enum (instead of enum class) so that it
can be implicitly casted to integers, since we want to use
optypes_enum2str[] for operation name printing.

witchel commented:
I'm not sure we want "raw" enums.  Doesn't our coding standard specify typed enums?  There are various ways to serialize typed enums without making the code unreadable.

https://stackoverflow.com/questions/8357240/how-to-automatically-convert-strongly-typed-enum-into-int

witchel commented:
Adding @ddn0 as a reviewer to weigh in on enum hygiene.

yige-hu committed:
Use class enum.

yige-hu committed:
Move enum DataTypes to my own header.

ddn0 commented:
Yup, `static_cast<int>(enum_class_value)` is the canonical way to get an int from an `enum class`. One characteristic of `enum class` over `enum` is avoiding implicit conversions from enum values to ints and vice versa, so we should expect direct conversions to look a little ugly, because they are to be avoided.

Here's an example of stringifying an `enum class`:
https://godbolt.org/z/7rqvad

Even if we did want to use `enum`, which I wouldn't recommend, `enum` values should be prefixed (e.g., `kOpLogInvalid`) as they introduce names into the containing scope.

yige-hu commented:
> Yup, `static_cast<int>(enum_class_value)` is the canonical way to get an int from an `enum class`. One characteristic of `enum class` over `enum` is avoiding implicit conversions from enum values to ints and vice versa, so we should expect direct conversions to look a little ugly, because they are to be avoided.
> 
> Here's an example of stringifying an `enum class`:
> https://godbolt.org/z/7rqvad
> 
> Even if we did want to use `enum`, which I wouldn't recommend, `enum` values should be prefixed (e.g., `kOpLogInvalid`) as they introduce names into the containing scope.

@ddn0 Yep, I changed OpTypes back to enum class. But I'm still using optypes_enum2str[] array for name printing (with static_cast on the enum class). Do you think it's better I put the string array here or my own header in the enterprise repo?

ddn0 commented on libgalois/include/galois/OpLog.h:
```diff
@@ -19,6 +19,19 @@ enum class OpTypes {
   kOpEdgePropVal,
 };
 
+const char* const optypes_enum2str[] = {
```
- Comment on the line above:

I think
```cpp
enum class OpType {
  kInvalid = 0,
  kNodeAdd,
  kNodeDel,
  ...
}; 
```

might be better. "Op" can be dropped from the values because the only way to refer to them is prefixed with "OpType" anyway. Starting from `= 0` is less surprising. What you wrote will result in the same enum values but so would `enum class { kInvalid, kOpNodeAdd, ....}`.

If these values are to be serialized, you can also consider explicitly numbering all the values to prevent any problems is someone accidentally reordered the values.

yige-hu replied:
This is not from this PR. It's some old code from @witchel . Or I can modify it in this PR instead, if you think it's better.

ddn0 replied:
I have a small preference for `OpType` and the `enum class` values not having "Op" in their names.

Whether this should be now or later is up to you.

ddn0 commented on libgalois/include/galois/OpLog.h:
```diff
@@ -19,6 +19,19 @@ enum class OpTypes {
+    "NodeDel",     /* kOpNodeDel */
+    "EdgeAdd",     /* kOpEdgeAdd */
+    "EdgeDel",     /* kOpEdgeDel */
+    "NodePropDel", /* kOpNodePropDel */
```
For code like this I'd recommend this pattern:
```
std::string OpTypeToString(OpType t) {
  switch (t) {
    case OpType::kInvalid:
      return "Invalid";
    ...
  }
}
```

it is more robust to reordering and newer compilers will warn you if you miss a case.

It also wouldn't hurt to move implementation details like the body of `OpTypeToString` to a separate .cpp file. Though that can be done later.

yige-hu replied:
I think using string arrays to translate enum-s to their correspondent names is quite common (everywhere in the Linux kernel, at least). People just put the two classes together and manually maintain their consistency. (Maybe it's not common for C++, though.) I can move optypes_enum2str[] to my local code, or use the switch-case instead (I think I'll follow your advice to use switch-case). If so, I'll delete this PR.

ddn0 replied:
I think for the current pattern to work, you'd also need to declare `optypes_enum2str[]` `static`; otherwise, we'd define `optypes_enum2str` in each translation unit that included this file and the linker might get angry. We are saved by our current linker policy which explicitly hides symbols not marked `GALOIS_EXPORT` but on macOS that particular feature is not enabled. Currently the macOS builds are failing for a different reason but this is something to be aware of.

> I think using string arrays to translate enum-s to their correspondent names is quite common 

Yeah, it definitely is a reasonable approach, though it is very C-like. C and C++ have many false friends [1]---patterns that make a lot of sense in one language but are out of place in another.

[1] https://en.wikipedia.org/wiki/False_friend

yige-hu replied:
I moved all the things to my local header files in the enterprise repository. I used switch-case instead of string array translation. I'll now close this PR. Thanks for all the help!

ddn0 commented on libgalois/include/galois/OpLog.h:
```diff
@@ -19,6 +19,19 @@ enum class OpTypes {
+    "EdgePropVal", /* kOpEdgePropVal */
+    "Consumed",    /* kConsumed */
+};
+
```
comment on the lines below: SetNProp => SetNodeProp. Most of the other code writes "node" and "edge" out, though this is more important for public symbols.

yige-hu replied:
This is not from this PR. It's some old code from @witchel . Or I can modify it in this PR instead, if you think it's better.

ddn0 replied:
I don't think these names are (easily) public so it doesn't bother me, but this is something to consider during future iterations.

witchel replied:
That name was my bad.  I was just seeing how it felt to shorten the identifiers.

In general, I think it is more straightforward to just spell out node/edge and have separate functions for them even though the temptation to have a single algorithm that can deal with both is powerful.

ddn0 replied:
I feel your pain.

I created this widget in PropertyFileGraph https://github.com/KatanaGraph/katana/blob/54da9508373e093a1184955fa23d5dbe12e0d7bd/libgalois/include/galois/graphs/PropertyFileGraph.h#L270
to allow library code to reuse as much between edges and nodes while still presenting separately named edge and node functions to the end-user.

witchel replied:
When collecting a set of pointers to member functions one must ask, "Am I a force for good?" 😎
